### PR TITLE
PPC support

### PR DESCRIPTION
**Pull request by simigo79**
I've made only few tests so far: basic arithmetic, and strlen(). No stack-related tests, no tests for resource leaks on multiple runs, no tests for hooks. Also it is limited to e500v2 and built only in ubuntu/amd64/gcc environment. So this is surely work in progress. 
However the changes to the common part of unicorn are few and easy to review and most likely they are not going to introduce any problem with non-ppc platforms. So if you accept the changes you are going to have  an otherwise  stable engine with experimental unstable ppc support.

### CODE REVIEW

**Comment by aquynh** [comment]
great work! let me and community review this before we can merge.

quick question: those "libdecnumber" code are really necessary?

**Comment by aquynh** [comment]
would you look at the Travis report?

**Comment by simigo79** [comment]
The libdecnumber is used somewhere in PPC code. I can try to get rid of it later.
The checks failed for three reasons as far as I can see: trivial one with no return value, some stuff with 'ffs' function  and I don't understand what is the third problem? Missing python sample?

**Comment by aquynh** [comment]
you may fix the bugs you found, and let the Travis try again. 

btw, the long list of warnings look worrisome. can you fix them too?

**Comment by simigo79** [comment]
Ok, I let you know when I fix the problems. 
I don't like those warnings as well. I think I can fix them but it doesn't seem to be the main problem at this moment.
About libdecnumbers, they are only used for DFP support present in IBM POWER6 and above. Me I don't need support for these cores (at least now). However removing it completely would make it harder to re-implement if someone needs them.

**Comment by aquynh** [comment]
yes lets keep libdecnumber.

**Comment by simigo79** [comment]
Hi, I tried to fix the problems I saw in Travis checks.

**Comment by aquynh** [comment]
very nice, but some tests still fail now?

https://travis-ci.org/github/unicorn-engine/unicorn/builds/685399803

**Comment by simigo79** [comment]
Yes, it looks like it expects python sample stdout to be exactly the same as C sample, does it? Also when the python sample test passes will it require ruby, pascal, whatever else sample?
Another problem is ffs() function. It seems I have to implement it myself in mingw.

**Comment by aquynh** [comment]
> Yes, it looks like it expects python sample stdout to be exactly the same as C sample, does it?

yes.

> Also when the python sample test passes will it require ruby, pascal, whatever else sample?

only Python check is performed on Travis

**Comment by simigo79** [comment]
Still Travis checks fail and I have no idea what's wrong with check 1959.10 It looks like a problem with test environment.

**Comment by simigo79** [comment]
Ok, now all Travis checks succeeded. Please review the changes to common files.
Once again, ppc support is very unstable, experimental, etc but it should not affect other platforms.

File: README.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+An attempt to add PPC support to Unicorn Engine.
```
Comment (aquynh) [inline README.md:1]: please remove these lines in README

**Comment by aquynh** [comment]
can you please update the CMakeLists.txt file, too?

**Comment by simigo79** [comment]
Fixed readme.
I can try to update CMakeLists.txt but I never dealt with it besides running cmake to build someone's project.

**Comment by simigo79** [comment]
Funny but fixing readme caused Travis check failure.

**Comment by aquynh** [comment]
ok i can update CMake

File: qemu/target-ppc/translate.c [hunk 2]
```diff
@@ -11409,6 +11409,14 @@ static inline void gen_intermediate_code_internal(PowerPCCPU *cpu,
         if (unlikely(qemu_loglevel_mask(CPU_LOG_TB_OP | CPU_LOG_TB_OP_OPT))) {
             tcg_gen_debug_insn_start(tcg_ctx, ctx.nip);
         }
+
+    // Unicorn: trace this instruction on request
```
Comment (aquynh) [inline qemu/target-ppc/translate.c:11413]: please indent with spaces, not tabs.

**Comment by simigo79** [comment]
I've done quite a lot of testing. Haven't found any error so far.

**Comment by aquynh** [comment]
very nice work! can you please fix the remaining compiler warnings of ppc?

**Comment by aquynh** [comment]
i tried `sample_ppc` on Ubuntu 18.04 64bit, and got this error:

```
$ ./samples/sample_ppc
Emulate PPC code
Failed on uc_open() with error returned: 2 (Invalid/unsupported architecture (UC_ERR_ARCH))
```

**Comment by simigo79** [comment]
This is strange. I just tested it in Mint 19.3 amd64 which is the same as Ubuntu 18.04 and it worked.
Are you sure the library in the library path is correct one with ppc support compiled? Could you try it with LD_LIBRARY_PATH explicitly pointing to the correct library?

**Comment by aquynh** [comment]
i tried other samples, like sample_arm, and they all work.

**Comment by aquynh** [comment]
sorry, that was my fault: it was linked against older version.

please fix the remaining warnings, then we will merge this, thanks.

**Comment by aquynh** [comment]
nice, only 2 more warnings left now: https://travis-ci.org/github/unicorn-engine/unicorn/jobs/691080586

**Comment by aquynh** [comment]
i found an issue: i ran "make header" (since you modified qemu/header_gen.py) then tried to build, and got the error below:

```
...
  GEN     libunicorn.so.1
qemu/ppc-softmmu/target-ppc/fpu_helper.o: In function `helper_fsqrt':
/home/me/projects/unicorn/qemu/target-ppc/fpu_helper.c:918: multiple definition of `helper_fsqrt'
qemu/x86_64-softmmu/target-i386/fpu_helper.o:/home/me/projects/unicorn/qemu/target-i386/fpu_helper.c:889: first defined here
qemu/ppc-softmmu/target-ppc/int_helper.o: In function `helper_divs':
/home/me/projects/unicorn/qemu/target-ppc/int_helper.c:332: multiple definition of `helper_divs'
qemu/m68k-softmmu/target-m68k/op_helper.o:/home/me/projects/unicorn/qemu/target-m68k/op_helper.c:201: first defined here
collect2: error: ld returned 1 exit status
Makefile:328: recipe for target 'libunicorn.so.1' failed
make: *** [libunicorn.so.1] Error 1
```

**Comment by aquynh** [comment]
again, thanks for this amazing work! i fixed some issues with header, made some cleanup, and merged this into the new "next" branch, to get more test. later, we will merge "next" to "master" for v1.0.3.

- please send pull req for the "next" branch for PPC from now on.
- please send pull req to add your name to CREDITS.TXT.

a question: what is missing for PPC64 support?

**Comment by aquynh** [comment]
i updated a comment on PPC32 mode: https://github.com/unicorn-engine/unicorn/commit/90ed828e90364e9105228979b90cd48ba202e16f

**Comment by aquynh** [comment]
i removed some unused files at https://github.com/unicorn-engine/unicorn/commit/f77bdb3c7ecc98b4d486b4c3004e647050e1d9d8

there are still a bunch of unused files in https://github.com/unicorn-engine/unicorn/tree/next/qemu/hw/ppc, please remove them.

**Comment by aquynh** [comment]
CMake now supports PPC.

however,  msvc/unicorn/ppc-softmmu/ is still missing, so Windows build fails. can you make a pull req for that?

**Comment by aquynh** [comment]
i opened a new issue regarding MSVC support: https://github.com/unicorn-engine/unicorn/issues/1269

**Comment by aquynh** [comment]
let me close this pull req, since it was already merged.

can you make pull req to the next branch instead.?

**Comment by simigo79** [comment]
As far as I remember there were some global name conflicts between ppc32 and ppc64, so I removed ppc64 from the build to speed up ppc32 development. Most likely the conflicts can be solved by updating header_gen.py 
Also only one ppc platform is implemented (e500 which is kind of "generic nxp") by explicitly initializing its types in generic ppc_cpu_register_types() function. For ppc64 one has to select a different platform (e500 is 32 bit) and replace e500 types with the ones for selected platform. Or better implement a way to select ppc platform. Also I replaced e500 mmu with "real" mmu (that is physical == virtual dummy). For the selected ppc64 platform one has to replace its mmu with "real" one as well. (I think that mmu emulation is impractical for unicorn).

I knew about unused files but I didn't remove them because some of them would be needed if someone wants to implement other flavors of ppc. 

Also there is pointer_fix.py script left in qemu/target-ppc directory. It is not needed.  Please remove it, it is one of the scripts I used to process qemu sources for unicorn.

**Pull request closed without merging**